oecophylla/taxonomy/metaphlan2.py

```diff
@@ -0,0 +1,287 @@
+                                 (line, filename))
+
+        f.close()
+        return dict
+    except IOError:
+        raise IOError('Cannot read file "%s"' % filename)
+
+
+def read_metaphlan_markers_info(filename):
+    """ Reads the MetaPhlAn markers_info.txt file.
+
+    MetaPhlAn's OTU analogous are 'clades'. Currently, they have around 8900.
+    A 'clade' is composed of one or many (sub)sequences of specific marker
+    genes. Those marker genes come from three sources: 1) genbank: "^gi|",
+    2) gene: "^GeneID:", and 3) NCBI nr: "^NC_".
+
+    Parameters
+    ----------
+    filename : str
+        Path to the filename 'markers_info' of MetaPhlAn.
+
+    Returns
+    -------
+    A dict with an entry for each 'clade'. Their values are dicts themselves,
+    with keys that refer to one of the three sequence sources. And their values
+    are sets of marker gene IDs. For example:
+    's__Escherichia_phage_vB_EcoP_G7C': {'GeneID': {'11117645', '11117646'}}
+
+    Raises
+    ------
+    IOError
+        If the file cannot be read.
+    """
+    clades = {}
+    try:
+        file = open(filename, 'r')
+        for line in file:
+            if line.startswith('gi|'):
+                type_ids = 'gi'
+                accession = (line.split('\t')[0]).split('|')[1]
+            elif line.startswith('GeneID:'):
+                type_ids = 'GeneID'
+                accession = (line.split('\t')[0]).split(':')[1]
+            elif line.startswith('NC_'):
+                type_ids = 'NC'
+                accession = line.split('\t')[0]
+            else:
+                type_ids = None
+                accession = None
+
+            if (type_ids is not None) and (accession is not None):
+                clade = line.split("clade': '")[1].split("'")[0]
+                if clade not in clades:
+                    clades[clade] = {}
+                if type_ids not in clades[clade]:
+                    clades[clade][type_ids] = {}
+                clades[clade][type_ids][accession] = True
+
+        for clade in clades:
+            for type_id in clades[clade]:
+                clades[clade][type_id] = set(clades[clade][type_id].keys())
+
+        file.close()
+        return clades
+
+    except IOError:
+        raise IOError('Cannot read file "%s"' % filename)
+
+
+def _read_ncbitaxonomy_file(filename):
+    """ A generic function to read an NCBI taxonomy file, which is delimited
+    by '\t|\t?'.
+
+    Parameters
+    ----------
+    filename : str
+        Path to a file from an NCBI taxonomy dump.
+
+    Returns
+    -------
+    A dict. Key = first field of file, Value = second field of file.
+
+    Raises
+    ------
+    IOError
+        If the file cannot be read.
+    ValueError
+        If IDs of entries cannot be converted into int.
+    """
+    entries = {}
+    try:
+        file = open(filename, 'r')
+        for line in file:
+            fields = list(map(str.strip, line.split('|')))
```

Comment:
Why not `fields = line.split('\t|\t')`?

Reply:
I tried your suggestion and it failed :-/

Reply:
Shouldn't have -- sure that it's tab-pipe-tab and not space-pipe-space?

```python
>>> a = 'foo | bar'
>>> a.split(' | ')
['foo', 'bar']
>>> a = 'foo\t|\tbar'
>>> a.split('\t|\t')
['foo', 'bar']
```

Reply:
I also tried \s+|\s+, which fails.
For the sake of time, I don't want to investigate in full detail, but I have the feeling that the files are not consistently formatted. Otherwise, I could have used pandas.read_csv for example.

Reply:
Sure, for future reference what you want is re.split:

```python
>>> import re
>>> a = 'foo | bar'
>>> b = 'foo   | bar'
>>> c = 'foo\t'| bar'
>>> c = 'foo\t| bar'
>>> re.split('\s+\|\s+', a)
['foo', 'bar']
>>> re.split('\s+\|\s+', b)
['foo', 'bar']
>>> re.split('\s+\|\s+', c)
['foo', 'bar']
```

Reply:
yes, that would be great to have. As a former Perl coder, I love regular expressions. Unfortunately, they are so damn slow in Python (even the compiled ones), that I refrained from using them in this language :-(

Reply:
I used to do @tanaes 's `re.split()` to handle taxdumps, then I switched to `line.split('\t|\t')` for the reason @sjanssen2 mentioned. But meanwhile, `line.split('\s+|\s+')` will not work because `+` is a regex symbol and is not supported by plain `split`.